### PR TITLE
[MAINTENANCE] Enable SparkDatasource add_asset() methods to accept "header" and "infer_schema" arguments; fix typos in docstrings.

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* Make sure that `add_<file_type>_asset()` methods of every `SparkDatasource` accepts `header` and `infer_schema` directives.
* Fix typos in `test_connection()` of cloud file storage `Datasource` classes to use `_get_<cloud_storage_provider>_client()` method as opposed to the reference to the `client` handle (so that if the `client` handle is `Non`, it can get initialized properly).
* Fix typos in docstrings.
 
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- JIRA: GREAT-1462/GREAT-1692/GREAT-1690/GREAT-1691/GREAT-1619
-
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: tests/datasource/fluent/test_pandas_filesystem_datasource.py [hunk 1]
```diff
@@ -295,7 +295,7 @@ def test_data_asset_reader_options_passthrough(
             )
             .add_csv_asset(
                 "my_csv",
-                batching_regex=r"yellow_tripdata_sample_(?P<year>\d{4})-(?P<month>\d{2}).csv",
+                batching_regex=r"yellow_tripdata_sample_(?P<year>\d{4})-(?P<month>\d{2})\.csv",
```
Comment (anthonyburdi) [inline tests/datasource/fluent/test_pandas_filesystem_datasource.py:298]: Why was this test passing without this change?

File: great_expectations/datasource/fluent/data_asset/data_connector/data_connector.py [hunk 2]
```diff
@@ -118,11 +117,30 @@ def get_data_reference_count(self) -> int:
...
+    @abstractmethod
+    def get_matched_data_reference_count(self) -> int:
+        """
+        This interface method returns number of all (e.g., cached) matched data references (useful for diagnostics).
+
+        Returns:
+            int -- number of data references identified
+        """
+        pass
+
```
Comment (anthonyburdi) [inline great_expectations/datasource/fluent/data_asset/data_connector/data_connector.py:139]: Are there tests for these new methods for each of the data connectors? I don't see them in this PR.

**Review by anthonyburdi — COMMENTED** [review]
I have a few questions, noted in the review. I would feel more comfortable if someone working more closely with ZEP could review to approve the rationale for the additional methods but otherwise makes sense to me aside from the comments.

File: tests/datasource/fluent/test_pandas_filesystem_datasource.py (hunk 1 above)
Comment (alexsherstinsky) [reply tests/datasource/fluent/test_pandas_filesystem_datasource.py:298]: @anthonyburdi This change makes the regular expression tighter -- and serves as a better example for anybody interested in learning from our examples.  Here, it is very clear that "\." is intended, while forgetting the backslash turns it into "any character" (just stating the obvious here for the sake of completeness).  The reason the tests pass is that the groups were captured regardless (the "month" group in particular, since it is near the period).  Thanks!

Comment (anthonyburdi) [reply tests/datasource/fluent/test_pandas_filesystem_datasource.py:298]: Thanks for clarifying!

File: great_expectations/datasource/fluent/data_asset/data_connector/data_connector.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/datasource/fluent/data_asset/data_connector/data_connector.py:139]: @anthonyburdi They do indeed appear in the `FilePathDataConnector` -- that level is sufficient for the subclasses.  Thanks.

Comment (anthonyburdi) [reply great_expectations/datasource/fluent/data_asset/data_connector/data_connector.py:139]: I'm still not seeing tests for these methods - am I missing them? I'm searching through the codebase for any tests that call these methods and coming up empty.

Comment (alexsherstinsky) [reply great_expectations/datasource/fluent/data_asset/data_connector/data_connector.py:139]: @anthonyburdi Are you seeing the implementations in `https://github.com/great-expectations/great_expectations/pull/7296/files#diff-d4fbdd20ca3f7bc62cb2b2178c5a5adcbd9edd3a2188e07f5d732df693150b1d` (one of the files changed)?  In terms of tests, they do not appear in the automated tests; I have used them in the Jupyter notebook (as part of UAT).  Thanks.

Comment (Kilo59) [reply great_expectations/datasource/fluent/data_asset/data_connector/data_connector.py:139]: Do these methods have to implemented?
Should they throw a `NotImplementedError`?

File: great_expectations/datasource/fluent/spark_azure_blob_storage_datasource.py [hunk 3]
```diff
@@ -120,6 +123,8 @@ def add_csv_asset(
         name: str,
         batching_regex: Union[re.Pattern, str],
         container: str,
+        header: bool = False,
+        infer_schema: bool = False,
```
Comment (Kilo59) [inline great_expectations/datasource/fluent/spark_azure_blob_storage_datasource.py:127]: `non-blocking`
This probably belongs in a different PR but...
I would expect that the `spark` `read_csv` equivalent also takes a lot more arguments than just these 2??

**Review by Kilo59 — APPROVED** [review]
Couple non-blocking questions.

File: great_expectations/datasource/fluent/spark_azure_blob_storage_datasource.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/datasource/fluent/spark_azure_blob_storage_datasource.py:127]: @Kilo59 Yes, I would like to collaborate with you on this -- as well as the following two items:
* More reader types for `SparkDatasource` (it is capable of much more than `CSV`);
* Making `add_<type>_asset()` for `SparkDatasource` more elegant.

Thank you.

File: great_expectations/datasource/fluent/data_asset/data_connector/data_connector.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/datasource/fluent/data_asset/data_connector/data_connector.py:139]: @Kilo59 I think that they should all be implemented (these implementations are very simple, and the methods are really valuable; in fact, some of them give us the `test_connection()` implementation, and also I am using them in UAT).  Thanks.
